README
==================
Pomodoro-Indicator
==================

Pomodoro-Indicator an indicator to work with the Pomodoro Technique

For more information:
https://en.wikipedia.org/wiki/Pomodoro_Technique

============
Requirements
============

Required dependencies:

    gir1.2-gtk-3.0,
    gir1.2-gdkpixbuf-2.0,
    gir1.2-gstreamer-1.0,
    gir1.2-gst-plugins-base-1.0,
    gir1.2-appindicator3-0.1,
    gir1.2-notify-0.7,
    python3,
    python3-gi,
    python3-dbus

========
Download
========

Download the code:

    $ git clone https://github.com/ImMaximus1/pomodoro-indicator.git

=======
Install
=======

    $ sudo python setup.py install


debian/changelog
pomodoro-indicator (0.2.0-0extras15.10.9) wily; urgency=low

  * Add Russian translation
  * Delete Spanish translation

 -- Max Medvedovsky <mmedvedovsky@gmail.com>  Fri, 04 Jan 2019 18:48:21 +0200

pomodoro-indicator (0.2.0-0extras15.10.6) wily; urgency=medium

  * Stop on breaks don't break set and switch to the next session
  * Stop on session don't break set and switch to the stopped session
  * Add notifications on break and sessions stop

 -- Max Medvedovsky <mmedvedovsky@gmail.com>  Tue, 12 Jun 2016 18:06:11 +0200

pomodoro-indicator (0.2.0-0extras15.10.3) wily; urgency=medium

  * Updated README
  * Fixed dependencies

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Wed, 01 Jun 2016 22:17:24 +0200

pomodoro-indicator (0.2.0-0extras15.10.0) wily; urgency=medium

  * Added more icons (60 images)
  * Session goes forward
  * Break goes backward
  * Added option to set number of pomodoros
  * Added long break after end max number of pomodoros

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Mon, 28 Mar 2016 23:37:32 +0200

pomodoro-indicator (0.1.0-0extras15.10.0) wily; urgency=medium

  * Upgraded to wily

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sun, 24 Jan 2016 10:14:30 +0100

pomodoro-indicator (0.1.0-0extras14.04.1) trusty; urgency=low

  * First release

 -- Lorenzo Carbonell <atareao@trusty>  Fri, 02 May 2014 10:40:01 +0200


msgfmt.py
# -*- coding: utf-8 -*-
# Written by Martin v. Lwis <loewis@informatik.hu-berlin.de>
# Plural forms support added by alexander smishlajev <alex@tycobka.lv>
"""
Generate binary message catalog from textual translation description.

This program converts a textual Uniforum-style message catalog (.po file) into
a binary GNU catalog (.mo file).  This is essentially the same function as the
GNU msgfmt program, however, it is a simpler implementation.

Usage: msgfmt.py [OPTIONS] filename.po

Options:
    -o file
    --output-file=file
        Specify the output file to write to.  If omitted, output will go to a
        file named filename.mo (based off the input file name).

    -h
    --help
        Print this message and exit.

    -V
    --version
        Display version information and exit.
"""

import sys
import os
import getopt
import struct
import array

__version__ = "1.1"

MESSAGES = {}


def usage (ecode, msg=''):
    """
    Print usage and msg and exit with given code.
    """
    print >> sys.stderr, __doc__
    if msg:
        print >> sys.stderr, msg
    sys.exit(ecode)


def add (msgid, transtr, fuzzy):
    """
    Add a non-fuzzy translation to the dictionary.
    """
    global MESSAGES
    if not fuzzy and transtr and not transtr.startswith('\0'):
        MESSAGES[msgid] = transtr


def generate ():
    """
    Return the generated output.
    """
    global MESSAGES
    keys = MESSAGES.keys()
    # the keys are sorted in the .mo file
    keys.sort()
    offsets = []
    ids = strs = ''
    for _id in keys:
        # For each string, we need size and file offset.  Each string is NUL
        # terminated; the NUL does not count into the size.
        offsets.append((len(ids), len(_id), len(strs), len(MESSAGES[_id])))
        ids += _id + '\0'
        strs += MESSAGES[_id] + '\0'
    output = ''
    # The header is 7 32-bit unsigned integers.  We don't use hash tables, so
    # the keys start right after the index tables.
    # translated string.
    keystart = 7*4+16*len(keys)
    # and the values start after the keys
    valuestart = keystart + len(ids)
    koffsets = []
    voffsets = []
    # The string table first has the list of keys, then the list of values.
    # Each entry has first the size of the string, then the file offset.
    for o1, l1, o2, l2 in offsets:
        koffsets += [l1, o1+keystart]
        voffsets += [l2, o2+valuestart]
    offsets = koffsets + voffsets
    output = struct.pack("Iiiiiii",
                         # 0x950412deL,       # Magic
                         0x950412de,       # Magic
                         0,                 # Version
                         len(keys),         # # of entries
                         7*4,               # start of key index
                         7*4+len(keys)*8,   # start of value index
                         0, 0)              # size and offset of hash table
    output += array.array("i", offsets).tostring()
    output += ids
    output += strs
    return output


def make (filename, outfile):
    ID = 1
    STR = 2
    global MESSAGES
    MESSAGES = {}

    # Compute .mo name from .po name and arguments
    if filename.endswith('.po'):
        infile = filename
    else:
        infile = filename + '.po'
    if outfile is None:
        outfile = os.path.splitext(infile)[0] + '.mo'

    try:
        lines = open(infile).readlines()
    except IOError(msg):
        print >> sys.stderr, msg
        sys.exit(1)

    section = None
    fuzzy = 0

    # Parse the catalog
    msgid = msgstr = ''
    lno = 0
    for l in lines:
        lno += 1
        # If we get a comment line after a msgstr, this is a new entry
        if l[0] == '#' and section == STR:
            add(msgid, msgstr, fuzzy)
            section = None
            fuzzy = 0
        # Record a fuzzy mark
        if l[:2] == '#,' and (l.find('fuzzy') >= 0):
            fuzzy = 1
        # Skip comments
        if l[0] == '#':
            continue
        # Start of msgid_plural section, separate from singular form with \0
        if l.startswith('msgid_plural'):
            msgid += '\0'
            l = l[12:]
        # Now we are in a msgid section, output previous section
        elif l.startswith('msgid'):
            if section == STR:
                add(msgid, msgstr, fuzzy)
            section = ID
            l = l[5:]
            msgid = msgstr = ''
        # Now we are in a msgstr section
        elif l.startswith('msgstr'):
            section = STR
            l = l[6:]
            # Check for plural forms
            if l.startswith('['):
                # Separate plural forms with \0
                if not l.startswith('[0]'):
                    msgstr += '\0'
                # Ignore the index - must come in sequence
                l = l[l.index(']') + 1:]
        # Skip empty lines
        l = l.strip()
        if not l:
            continue
        # XXX: Does this always follow Python escape semantics?
        l = eval(l)
        if section == ID:
            msgid += l
        elif section == STR:
            msgstr += l
        else:
            print >> sys.stderr, 'Syntax error on %s:%d' % (infile, lno), \
                  'before:'
            print >> sys.stderr, l
            sys.exit(1)
    # Add last entry
    if section == STR:
        add(msgid, msgstr, fuzzy)

    # Compute output
    output = generate()

    try:
        open(outfile,"wb").write(output)
    except IOError(msg):
        print >> sys.stderr, msg


def main ():
    try:
        opts, args = getopt.getopt(sys.argv[1:], 'hVo:',
                                   ['help', 'version', 'output-file='])
    except getopt.error(msg):
        usage(1, msg)

    outfile = None
    # parse options
    for opt, arg in opts:
        if opt in ('-h', '--help'):
            usage(0)
        elif opt in ('-V', '--version'):
            print >> sys.stderr, "msgfmt.py", __version__
            sys.exit(0)
        elif opt in ('-o', '--output-file'):
            outfile = arg
    # do it
    if not args:
        print >> sys.stderr, 'No input file given'
        print >> sys.stderr, "Try `msgfmt --help' for more information."
        return

    for filename in args:
        make(filename, outfile)


if __name__ == '__main__':
    main()


src/comun.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# comun.py
#
# This file is part of Pomodoro-Indicator
#
# Copyright (C) 2014
# Lorenzo Carbonell Cerezo <lorenzo.carbonell.cerezo@gmail.com>
# Max Medvedovsky <mmedvedovsky@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import os
import locale
import gettext

__author__ = 'Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>'
__copyright__ = 'Copyright (c) 2014 Lorenzo Carbonell'
__license__ = 'GPLV3'
__url__ = 'http://www.atareao.es'


######################################

def is_package():
    return not os.path.dirname(os.path.abspath(__file__)).endswith('src')

######################################


APP = 'pomodoro-indicator'
APPCONF = APP + '.conf'
APPDATA = APP + '.data'
APPNAME = 'Pomodoro-Indicator'
CONFIG_DIR = os.path.join(os.path.expanduser('~'), '.config')
CONFIG_APP_DIR = os.path.join(CONFIG_DIR, APP)
CONFIG_FILE = os.path.join(CONFIG_APP_DIR, APPCONF)
DATA_FILE = os.path.join(CONFIG_APP_DIR, APPDATA)
AUTOSTART_DIR = os.path.join(CONFIG_DIR, 'autostart')
# FILE_AUTO_START = os.path.join(AUTOSTART_DIR,
#                                'pomodoro-indicator-autostart.desktop')
FILE_AUTO_START = os.path.join(AUTOSTART_DIR,
                               'extras-pomodoro-indicator.desktop')
if not os.path.exists(CONFIG_APP_DIR):
    os.makedirs(CONFIG_APP_DIR)

print(os.path.dirname(os.path.abspath(__file__)))

if is_package():
    ROOTDIR = '/opt/extras.ubuntu.com/pomodoro-indicator/share/'
    LANGDIR = os.path.join(ROOTDIR, 'locale-langpack')
    APPDIR = os.path.join(ROOTDIR, APP)
    ICONDIR = os.path.join(APPDIR, 'icons')
    # SOCIALDIR = os.path.join(APPDIR, 'social')
    SOUNDIR = os.path.join(APPDIR, 'sounds')
    CHANGELOG = os.path.join(APPDIR, 'changelog')
    # FILE_AUTO_START_ORIG = os.path.join(APPDIR,
    #                                     'pomodoro-indicator-autostart.desktop')
    FILE_AUTO_START_ORIG = os.path.join(APPDIR,
                                        'extras-pomodoro-indicator.desktop')
else:
    ROOTDIR = os.path.dirname(__file__)
    LANGDIR = os.path.join(ROOTDIR, 'template1')
    APPDIR = os.path.join(ROOTDIR, APP)
    ICONDIR = os.path.normpath(os.path.join(ROOTDIR, '../data/icons'))
    # SOCIALDIR = os.path.normpath(os.path.join(ROOTDIR, '../data/social'))
    SOUNDIR = os.path.normpath(os.path.join(ROOTDIR, '../data/sounds'))
    DEBIANDIR = os.path.normpath(os.path.join(ROOTDIR, '../debian'))
    CHANGELOG = os.path.join(DEBIANDIR, 'changelog')
    # FILE_AUTO_START_ORIG = os.path.join(os.path.normpath(os.path.join(
    #     ROOTDIR, '../data')), 'pomodoro-indicator-autostart.desktop')
    FILE_AUTO_START_ORIG = os.path.join(os.path.normpath(os.path.join(
        ROOTDIR, '../data')), 'extras-pomodoro-indicator.desktop')
ICON = os.path.join(ICONDIR, 'pomodoro-indicator.svg')
DEFAULT_SESSION_SOUND_FILE = os.path.join(SOUNDIR, 'ringbell.ogg')
DEFAULT_BREAK_SOUND_FILE = os.path.join(SOUNDIR, 'gong.ogg')
# Sound file for a long break start
DEFAULT_LONG_BREAK_SOUND_FILE = os.path.join(SOUNDIR, 'festival.ogg')
STATUS_ICON = {}
STATUS_ICON['light'] = (os.path.join(ICONDIR, 'pomodoro-start-light.svg'),
                        os.path.join(ICONDIR, 'pomodoro-stop-light.svg'))
STATUS_ICON['dark'] = (os.path.join(ICONDIR, 'pomodoro-start-dark.svg'),
                       os.path.join(ICONDIR, 'pomodoro-stop-dark.svg'))

PARAMS = {
    'first-time': True,
    'version': '',
    'autostart': True,
    'theme': 'dark',
    'break_theme': 'light',  # Set type of break's icons like an inversion of selected team
    'session_length': 25,
    'break_length': 5,
    'long_break_length': 20,
    'number_of_pomodoros': 4,
    'play_sounds': True,
    # 'session_sound_file': 'default',
    'session_sound_file': DEFAULT_SESSION_SOUND_FILE,
    # 'break_sound_file': 'default'
    'break_sound_file': DEFAULT_BREAK_SOUND_FILE,
    'long_break_sound_file': DEFAULT_LONG_BREAK_SOUND_FILE
}

f = open(CHANGELOG, 'r')
line = f.readline()
f.close()
pos = line.find('(')
posf = line.find(')', pos)
VERSION = line[pos+1:posf].strip()
if not is_package():
    VERSION = VERSION + '-src'
try:
    current_locale, encoding = locale.getdefaultlocale()
    language = gettext.translation(APP, LANGDIR, [current_locale])
    language.install()
    _ = language.gettext
except Exception as e:
    _ = str


src/pomodoro_indicator.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# pomodoro-indicator.py
#
# This file is part of Pomodoro-Indicator
#
# Copyright (C) 2014
# Lorenzo Carbonell Cerezo <lorenzo.carbonell.cerezo@gmail.com>
# Max Medvedovsky <mmedvedovsky@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import gi
try:
    gi.require_version('Gtk', '3.0')
    gi.require_version('Gst', '1.0')
    gi.require_version('AppIndicator3', '0.1')
    gi.require_version('Notify', '0.7')
except Exception as e:
    print(e)
    exit(1)
from gi.repository import Gtk
from gi.repository import Gst
from gi.repository import GLib
from gi.repository import GdkPixbuf
from gi.repository import AppIndicator3 as appindicator
from gi.repository import Notify
from gi.repository import GObject
import os
import webbrowser
import dbus
from configurator import Configuration
from preferences_dialog import PreferencesDialog
from comun import _
import comun

gi.require_version('Gtk', '3.0')
gi.require_version('Gst', '1.0')

BUS_NAME = 'es.atareao.pomodoro'
BUS_PATH = '/es/atareao/pomodoro'
TOTAL_FRAMES = 60


def add2menu(menu, text=None, icon=None, conector_event=None,
             conector_action=None):
    if text is not None:
        menu_item = Gtk.ImageMenuItem.new_with_label(text)
        if icon:
            image = Gtk.Image.new_from_file(icon)
            menu_item.set_image(image)
            menu_item.set_always_show_image(True)
    else:
        if icon is None:
            menu_item = Gtk.SeparatorMenuItem()
        else:
            menu_item = Gtk.ImageMenuItem.new_from_file(icon)
            menu_item.set_always_show_image(True)
    if conector_event is not None and conector_action is not None:
        menu_item.connect(conector_event, conector_action)
    menu_item.show()
    menu.append(menu_item)
    return menu_item


class Pomodoro_Indicator(GObject.GObject):
    __gsignals__ = {
        'session_end': (GObject.SIGNAL_RUN_FIRST, GObject.TYPE_NONE, ()),
        'break_end': (GObject.SIGNAL_RUN_FIRST, GObject.TYPE_NONE, ()),
    }

    def __init__(self):
        GObject.GObject.__init__(self)
        self.pw = 0
        self.player = Gst.ElementFactory.make("playbin", "player")
        self.player.connect("about-to-finish",  self.on_player_finished)
        bus = self.player.get_bus()
        bus.connect("message", self.on_player_message)
        self.icon = comun.ICON
        self.active_icon = None
        self.about_dialog = None
        self.active = False
        self.animate = False
        self.frame = 0
        self.pomodoros = 0
        # Set or session trigger - what has been stopped
        self.isSet = False
        # Break or session trigger - what has been stopped
        self.isBreak = False
        self.notification = Notify.Notification.new('', '', None)
        self.read_preferences()
        self.indicator = appindicator.Indicator.new('Pomodoro-Indicator',
                                                    self.active_icon,
                                                    appindicator.
                                                    IndicatorCategory.
                                                    HARDWARE)
        self.indicator.set_status(appindicator.IndicatorStatus.ACTIVE)
        self.indicator.connect('scroll-event', self.on_scroll)

        menu = self.get_menu()
        self.indicator.set_menu(menu)
        self.connect('session_end', self.on_session_end)
        self.connect('break_end', self.on_break_end)

    def on_scroll(self, widget, steps, direction):
        self.on_pomodoro_start(None)

    def play(self, afile):
        self.player.set_property('uri', 'file://'+afile)
        self.player.set_state(Gst.State.PLAYING)

    def on_player_finished(self, player):
        self.player = None
        self.player = Gst.ElementFactory.make("playbin", "player")
        self.player.connect("about-to-finish",  self.on_player_finished)
        bus = self.player.get_bus()
        bus.connect("message", self.on_player_message)

    def on_player_message(self, bus, message):
        t = message.type
        if t == Gst.Message.EOS:
            self.player.set_state(Gst.State.NULL)
        elif t == Gst.Message.ERROR:
            self.player.set_state(Gst.State.NULL)
            err, debug = message.parse_error()
            print("Error: %s" % err, debug)

    # ################# main functions ####################

    def read_preferences(self):
        configuration = Configuration()
        self.first_time = configuration.get('first-time')
        self.version = configuration.get('version')
        self.theme = configuration.get('theme')
        self.break_theme = configuration.get('break_theme')
        self.max_pomodoros = int(configuration.get('number_of_pomodoros'))
        self.session_length = configuration.get('session_length')
        self.break_length = configuration.get('break_length')
        self.long_break_length = configuration.get('long_break_length')
        self.play_sounds = configuration.get('play_sounds')
        self.session_sound_file = configuration.get('session_sound_file')
        self.break_sound_file = configuration.get('break_sound_file')
        # Add different sound file for a long break start
        self.long_break_sound_file = configuration.get('long_break_sound_file')
        self.active_icon = comun.STATUS_ICON[configuration.get('theme')][0]

    # ################## menu creation ######################

    def get_help_menu(self):
        help_menu = Gtk.Menu()
        #
        homepage_item = Gtk.MenuItem(label=_(
            'Homepage'))
        homepage_item.connect('activate',
                              lambda x: webbrowser.open(
                                'http://www.atareao.es/'))
        homepage_item.show()
        help_menu.append(homepage_item)
        #
        help_item = Gtk.MenuItem(label=_(
            'Get help online...'))
        help_item.connect('activate',
                          lambda x: webbrowser.open(
                            'http://www.atareao.es/apps/la-tecnica-pomodoro-en\
-ubuntu-con-pomodoro-indicator/'))
        help_item.show()
        help_menu.append(help_item)
        #
        translate_item = Gtk.MenuItem(label=_(
            'Translate this application...'))
        translate_item.connect('activate',
                               lambda x: webbrowser.open(
                                'http://www.atareao.es/apps/la-tecnica-\
pomodoro-en-ubuntu-con-pomodoro-indicator/'))
        translate_item.show()
        help_menu.append(translate_item)
        #
        bug_item = Gtk.MenuItem(label=_(
            'Report a bug...'))
        bug_item.connect('activate',
                         lambda x: webbrowser.open(
                            'https://github.com/atareao/pomodoro-indicator/\
issues'))
        bug_item.show()
        help_menu.append(bug_item)
        #
        separator = Gtk.SeparatorMenuItem()
        separator.show()
        help_menu.append(separator)
        #
        twitter_item = Gtk.MenuItem(label=_(
            'Follow me in Twitter'))
        twitter_item.connect('activate',
                             lambda x: webbrowser.open(
                                'https://twitter.com/atareao'))
        twitter_item.show()
        help_menu.append(twitter_item)
        #
        googleplus_item = Gtk.MenuItem(label=_(
            'Follow me in Google+'))
        googleplus_item.connect('activate',
                                lambda x: webbrowser.open(
                                    'https://plus.google.com/\
118214486317320563625/posts'))
        googleplus_item.show()
        help_menu.append(googleplus_item)
        #
        facebook_item = Gtk.MenuItem(label=_(
            'Follow me in Facebook'))
        facebook_item.connect('activate',
                              lambda x: webbrowser.open(
                                'http://www.facebook.com/elatareao'))
        facebook_item.show()
        help_menu.append(facebook_item)
        #
        about_item = Gtk.MenuItem.new_with_label(_('About'))
        about_item.connect('activate', self.on_about_item)
        about_item.show()
        separator = Gtk.SeparatorMenuItem()
        separator.show()
        help_menu.append(separator)
        help_menu.append(about_item)
        #
        help_menu.show()
        return help_menu

    def get_menu(self):
        """Create and populate the menu."""
        menu = Gtk.Menu()

        self.pomodoro_start = Gtk.MenuItem.new_with_label(_('Start'))
        self.pomodoro_start.connect('activate', self.on_pomodoro_start)
        self.pomodoro_start.show()
        menu.append(self.pomodoro_start)

        self.pomodoro_reset = Gtk.MenuItem.new_with_label(_('Reset Pomodoro Set'))
        self.pomodoro_reset.connect('activate', self.on_pomodoro_reset)
        self.pomodoro_reset.show()
        menu.append(self.pomodoro_reset)

        separator1 = Gtk.SeparatorMenuItem()
        separator1.show()
        menu.append(separator1)
        #
        menu_preferences = Gtk.MenuItem.new_with_label(_('Preferences'))
        menu_preferences.connect('activate', self.on_preferences_item)
        menu_preferences.show()
        menu.append(menu_preferences)

        menu_help = Gtk.MenuItem.new_with_label(_('Help'))
        menu_help.set_submenu(self.get_help_menu())
        menu_help.show()
        menu.append(menu_help)
        #
        separator2 = Gtk.SeparatorMenuItem()
        separator2.show()
        menu.append(separator2)
        #
        menu_exit = Gtk.MenuItem.new_with_label(_('Exit'))
        menu_exit.connect('activate', self.on_quit_item)
        menu_exit.show()
        menu.append(menu_exit)
        #
        menu.show()
        return menu

    def on_pomodoro_start(self, widget):
        if not self.active:
            self.active = True
            self.pomodoro_start.set_label(_('Stop'))
            print('Session {0} of Pomodoro set starts'.format(self.pomodoros + 1))
            self.notification.update('Pomodoro-Indicator',
                                     _('Session {0} of Pomodoro set starts'.format(self.pomodoros + 1)),
                                     os.path.join(comun.ICONDIR,
                                                  'pomodoro-start-%s.svg' % (self.theme)))
            self.notification.show()
            self.countdown_session()
            # Stop trigger - what will has been stopped: break or session
            self.isBreak = False

            interval = int(self.session_length * 60 / TOTAL_FRAMES)
            self.start_working_process(interval, self.countdown_session)
        else:
            self.stop_working_process()
            self.active = False
            # If first session has just been stopped we will set pomodoro counter (self.pomodoros) to nil
            if self.pomodoros == 0 or self.pomodoros == self.max_pomodoros:
                self.pomodoros = 0
                self.isSet = True
                icon = os.path.join(comun.ICONDIR, 'pomodoro-indicator-%s-%s.svg' % (self.theme, 'set'))
            else:
                self.isSet = False
                icon = os.path.join(comun.ICONDIR, 'pomodoro-stop-%s.svg' % (self.theme))
            self.frame = 0
            self.pomodoro_start.set_label(_('Start'))
            self.indicator.set_icon(icon)
            print('Pomodoro {0} stop'.format("set" if self.isSet else "session {0[0]} on {0[1]}".format(
                    (self.pomodoros, "break") if self.isBreak else (self.pomodoros + 1, "session"))))
            if self.isSet:
                notification_message = _('Pomodoro set stop')
            else:
                if self.isBreak:
                    notification_message = _('Pomodoro session {0} on break stop').format(self.pomodoros)
                else:
                    notification_message = _('Pomodoro session {0} on session stop').format(self.pomodoros + 1)
            self.notification.update('Pomodoro-Indicator', notification_message, icon)
            self.notification.show()

    def on_pomodoro_reset(self, widget):
        self.stop_working_process()
        self.active = False
        self.pomodoros = 0
        # Set or session trigger - what has been stopped
        self.isSet = False
        # Break or session trigger - what has been stopped
        self.isBreak = False
        icon = os.path.join(comun.ICONDIR, 'pomodoro-start-%s.svg' % (self.theme))
        self.frame = 0
        self.pomodoro_start.set_label(_('Start'))
        self.indicator.set_icon(icon)
        print('Pomodoro Set reset')
        self.notification.update('Pomodoro-Indicator', _('Pomodoro Set reset'), icon)
        self.notification.show()

    def stop_working_process(self):
        if self.pw > 0:
            GLib.source_remove(self.pw)
            self.pw = 0

    def start_working_process(self, interval, countdown):
        if self.pw > 0:
            GLib.source_remove(self.pw)
        print('interval', interval)
        print('pomodoros', self.pomodoros)
        self.pw = GLib.timeout_add_seconds(interval, countdown)

    def on_break_end(self, widget):
        if self.pomodoros < self.max_pomodoros:
            self.frame = 0
            icon = os.path.join(comun.ICONDIR,
                                'pomodoro-indicator-%s-%02d.svg' % (self.theme,
                                                                    self.frame))
            self.notification.update('Pomodoro-Indicator', _('Break ends'), icon)
            self.notification.show()
            if self.play_sounds:
                self.play(self.break_sound_file)
            print('**************************')
            print(icon)
            self.indicator.set_icon(icon)

            self.notification.update('Pomodoro-Indicator',
                                     _('Session {0} starts'.format(self.pomodoros + 1)), icon)
            self.notification.show()
            # Stop trigger - what will has been stopped: break or session
            self.isBreak = False

            interval = int(self.session_length * 60 / TOTAL_FRAMES)
            self.start_working_process(interval, self.countdown_session)
        else:
            self.stop_working_process()
            self.active = False
            self.pomodoros = 0
            self.frame = 0
            self.pomodoro_start.set_label(_('Start'))
            # Indicate the end of pomodoro set
            icon = os.path.join(comun.ICONDIR, 'pomodoro-indicator-%s-%s.svg' % (self.theme, 'set'))
            self.indicator.set_icon(icon)
            self.notification.update('Pomodoro-Indicator', _('The Pomodoro set ends'), icon)
            self.notification.show()
            if self.play_sounds:
                self.play(self.break_sound_file)
            print('**************************')
            print(icon)

    def on_session_end(self, widget):
        self.active = True
        # One more pomodoro session has finished
        self.pomodoros += 1
        self.frame = TOTAL_FRAMES
        if self.pomodoros == self.max_pomodoros and self.max_pomodoros != 1:
            interval = int(self.long_break_length * 60 / TOTAL_FRAMES)
            message = _('Session {0} ends - long break starts'.format(self.pomodoros))
            # Add different sound file for a long break start
            sound_file = self.long_break_sound_file
        else:
            interval = int(self.break_length * 60 / TOTAL_FRAMES)
            message = _('Session {0} ends - break starts'.format(self.pomodoros))
            # Add different sound file for a long break start
            sound_file = self.session_sound_file
        # Set type of break's icons like an inversion of selected team
        icon = os.path.join(comun.ICONDIR,
                            'pomodoro-indicator-%s-%02d.svg' % (
                                 self.break_theme, self.frame))
        self.notification.update('Pomodoro-Indicator', message, icon)
        self.notification.show()
        # Add different sound file for a long break start
        if self.play_sounds:
            self.play(sound_file)
        self.indicator.set_icon(icon)
        # Stop trigger - what will been stopped: break or session
        self.isBreak = True

        self.countdown_break()
        self.start_working_process(interval, self.countdown_break)

    def countdown_session(self):
        if not self.active:
            self.animate = False
            icon = os.path.join(comun.ICONDIR,
                                'pomodoro-indicator-%s-%02d.svg' % (
                                    self.theme, self.frame))
            self.indicator.set_icon(icon)
            self.frame = 0
            return False
        else:
            self.animate = True
            icon = os.path.join(comun.ICONDIR,
                                'pomodoro-indicator-%s-%02d.svg' % (
                                    self.theme, self.frame))
            self.indicator.set_icon(icon)
            print('countdown_session', self.frame)
            print(icon)
            self.frame += 1
            if self.frame == TOTAL_FRAMES + 1:
                self.frame = TOTAL_FRAMES
                self.emit('session_end')
                return False
            return True

    def countdown_break(self):
        if not self.active:
            self.animate = False
            # Set type of break's icons like an inversion of selected team
            icon = os.path.join(comun.ICONDIR,
                                'pomodoro-indicator-%s-%02d.svg' % (
                                    self.break_theme, self.frame))
            self.indicator.set_icon(icon)
            self.frame = 0
            return False
        else:
            self.animate = True
            # Set type of break's icons like an inversion of selected team
            icon = os.path.join(comun.ICONDIR,
                                'pomodoro-indicator-%s-%02d.svg' % (
                                    self.break_theme, self.frame))
            self.indicator.set_icon(icon)
            print('countdown_break', self.frame)
            print(icon)
            self.frame -= 1
            if self.frame == -1:
                self.frame = 0
                self.emit('break_end')
                return False
            return True

    def get_about_dialog(self):
        """Create and populate the about dialog."""
        about_dialog = Gtk.AboutDialog()
        about_dialog.set_name(comun.APPNAME)
        about_dialog.set_version(comun.VERSION)
        about_dialog.set_copyright(
            'Copyrignt (c) 2014-2016\nLorenzo Carbonell Cerezo\nMax Medvedovsky')
        about_dialog.set_comments(_('An indicator for Pomodoro Technique'))
        about_dialog.set_license('''
This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <http://www.gnu.org/licenses/>.
''')
        about_dialog.set_website('http://www.atareao.es')
        about_dialog.set_website_label('http://www.atareao.es')
        about_dialog.set_authors([
            'Lorenzo Carbonell <https://launchpad.net/~lorenzo-carbonell>',
            'Max Medvedovsky <mmedvedovsky@gmail.com>',
            _('Sound:'),
            'Noah McGrath <https://twitter.com/Noah_McGrath>',
        ])
        about_dialog.set_documenters([
            'Lorenzo Carbonell <https://launchpad.net/~lorenzo-carbonell>',
        ])
        about_dialog.set_translator_credits(_('translator-credits'))
        about_dialog.set_icon(GdkPixbuf.Pixbuf.new_from_file(comun.ICON))
        about_dialog.set_logo(GdkPixbuf.Pixbuf.new_from_file(comun.ICON))
        about_dialog.set_program_name(comun.APPNAME)
        return about_dialog

    # ##################### callbacks for the menu #######################
    def on_preferences_item(self, widget, data=None):
        widget.set_sensitive(False)
        preferences_dialog = PreferencesDialog()
        if preferences_dialog.run() == Gtk.ResponseType.ACCEPT:
            preferences_dialog.close_ok()
            self.read_preferences()
        preferences_dialog.hide()
        preferences_dialog.destroy()
        self.indicator.set_icon(self.active_icon)
        widget.set_sensitive(True)

    def on_quit_item(self, widget, data=None):
        exit(0)

    def on_about_item(self, widget, data=None):
        if self.about_dialog:
            self.about_dialog.present()
        else:
            self.about_dialog = self.get_about_dialog()
            self.about_dialog.run()
            self.about_dialog.destroy()
            self.about_dialog = None

#################################################################


def main():

    if dbus.SessionBus().\
                          request_name('es.atareao.PomodoroIndicator') != \
                          dbus.bus.REQUEST_NAME_REPLY_PRIMARY_OWNER:
        print("application already running")
        exit(0)
    GObject.threads_init()
    Gst.init(None)
    Gst.init_check(None)
    Notify.init('pomodoro-indicator')
    pomodoro_indicator = Pomodoro_Indicator()
    Gtk.main()

if __name__ == "__main__":
    main()
